src/emqx_misc.erl

```diff
@@ -330,6 +331,39 @@ hexchar2int(I) when I >= $0 andalso I =< $9 -> I - $0;
 hexchar2int(I) when I >= $A andalso I =< $F -> I - $A + 10;
 hexchar2int(I) when I >= $a andalso I =< $f -> I - $a + 10.
 
+
+-spec maybe_middleman(module(), atom(), list()) -> term().
+maybe_middleman(Mod, Fun, Args) ->
+    [{message_queue_len, MQL}] = process_info(self(), [message_queue_len]),
+    MaxMQL = 10,
+    if MQL >= MaxMQL ->
+            with_middleman(Mod, Fun, Args);
+       true ->
+            apply(Mod, Fun, Args)
+    end.
+
+-spec with_middleman(module(), atom(), list()) -> term().
+with_middleman(Mod, Fun, Args) ->
+    Ref = make_ref(),
+    Parent = self(),
+    spawn_link(fun() ->
+                       Result = try apply(Mod, Fun, Args) of
+                                    R -> {ok, R}
+                                catch
+                                    EC:Err ->
+                                        {EC, Err}
+                                end,
+                       Parent ! {Ref, Result}
+               end),
+    receive
+        {Ref, Result} ->
+            case Result of
+                {ok, R} -> R;
+                {EC, Err} ->
+                    erlang:EC(Err)
```

Comment:
Raise with original stacktrace?

Reply:
How do I do that?

Reply:
Maybe [`erlang:raise/3`](https://www.erlang.org/doc/man/erlang.html#raise-3) can do that.

Reply:
Thanks, I didn't know about this function. However, mnesia only exits, which doesn't produce a stacktrace under normal conditions (unless you catch). So in this case it's probably unnecessary.

Reply:
yes, catch and return at line 354, and call `erlang:raise/3` here

```
                                    EC:Err:Stack ->
                                        {EC, Err, Stack}
```

Reply:
ah. if the stack is not interesting anyway. then ignore this.